seqfold/fold.go

```diff
@@ -0,0 +1,875 @@
+		ij1 := branches[abs((index-1)%len(branches))]
+		j1 := ij1.J
+		ij3 := branches[abs((index+1)%len(branches))]
+		i3, j3 := ij3.I, ij3.J
+
+		// add energy from unpaired bp to the right
+		// of the helix as though it was a dangling end
+		// if there's only one bp, it goes to whichever
+		// helix (this or the next) has the more favorable energy
+		unpaired_left := 0
+		unpaired_right := 0
+		de := 0.0
+		if index == len(branches)-1 && !helix {
+			// TODO what here?
+			// pass
+		} else if ij3 == ij {
+			unpaired_left = i2 - j1 - 1
+			unpaired_right = j3 - j2 - 1
+
+			if unpaired_left != 0 && unpaired_right != 0 {
+				de = Stack(i2-1, i2, j2+1, j2, fc)
+			} else if unpaired_right != 0 {
+				de = Stack(-1, i2, j2+1, j2, fc)
+				if unpaired_right == 1 {
+					de = min(Stack(i3, -1, j3, j3-1, fc), de)
+				}
+			}
+		} else if ij2 == ij {
+			unpaired_left = j2 - j1 - 1
+			unpaired_right = i3 - i2 - 1
+
+			if unpaired_left != 0 && unpaired_right != 0 {
+				de = Stack(i2-1, i2, j2+1, j2, fc)
+			} else if unpaired_right != 0 {
+				de = Stack(i2, i2+1, j2, -1, fc)
+				if unpaired_right == 1 {
+					de = min(Stack(i3-1, i3, -1, j3, fc), de)
+				}
+			}
+		} else {
+			unpaired_left = i2 - j1 - 1
+			unpaired_right = i3 - j2 - 1
+
+			if unpaired_left != 0 && unpaired_right != 0 {
+				de = Stack(i2-1, i2, j2+1, j2, fc)
+			} else if unpaired_right != 0 {
+				de = Stack(-1, i2, j2+1, j2, fc)
+				if unpaired_right == 1 {
+					de = min(Stack(i2-1, i2, j2+1, j2, fc), de)
+				}
+			}
+		}
+		e_sum += de
+		unpaired += unpaired_right
+		// assert unpaired_right >= 0
+		// TODO handle assertion
+		if unpaired_right < 0 {
+			panic("unpaired_right >= 0")
+		}
+
+		if ij2 != ij { // add energy
+			e_sum += W(i2, j2, fc).E
+		}
+
+	}
+
+	// assert unpaired >= 0
+	// TODO handle assertion
+	if unpaired < 0 {
+		panic("assert unpaired >= 0")
+	}
+
+	// penalty for unmatched bp and multi-branch
+	a, b, c, d := fc.Energies.MULTIBRANCH.A, fc.Energies.MULTIBRANCH.B, fc.Energies.MULTIBRANCH.C, fc.Energies.MULTIBRANCH.D
+	e_multibranch := a + b*float64(len(branches)) + c*float64(unpaired)
+
+	if unpaired == 0 {
+		e_multibranch = a + d
+	}
+
+	// energy of min-energy neighbors
+	e := e_multibranch + e_sum
+
+	// pointer to next structures
+	if helix {
+		// branches.pop()
+		branches = branches[:len(branches)-1]
+	}
+
+	return NucleicAcidStructure{E: e, Desc: fmt.Sprintf("BIFURCATION:%dn/%dh", unpaired, branches_count), Inner: branches}
+}
+
+// InternalLoop calculates the free energy of an internal loop.
+//
+// The first and last bp of both left and right sequences
+// are not themselves parts of the loop, but are the terminal
+// bp on either side of it. They are needed for when there's
+// a single internal looping bp (where just the mismatching
+// free energies are used)
+// Note that both left and right sequences are in 5' to 3' direction
+// This is adapted from the "Internal Loops" section of SantaLucia/Hicks, 2004
+// Args:
```

Comment:
rename variables and args based on this mapping for InternalLoop

Reply:
I did a pass, I'll do another later this week when I should have some more time.